### PR TITLE
🌱 fallback on build if kindest images are missing

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR fixes a problem in our E2E test, where test will fail if the kindest/node image for the requested Kubernetes version is not published yet.

With this PR instead we are changing our prePull sequence, and if a kindest image is not available on gitHub, we build it locally.

As a niece side effect, we are also dropping the BUILD_NODE_IMAGE_TAG env variable for our tests configuration...

/assign @srm09

### CODE REVIEW

File: scripts/ci-e2e-lib.sh [hunk 1]
```diff
@@ -40,27 +40,35 @@ capi:buildDockerImages () {
...
-# and resolves kubernetes version labels (e.g. latest) to the corresponding version numbers.
-k8s::resolveAllVersions() {
+# k8s::prepareKindestImages checks all the e2e test variables representing a Kubernetes version,
+# and makes sure a corresponding kindest/node image is available locally.
+k8s::prepareKindestImages() {
   if [ -n "${KUBERNETES_VERSION:-}" ]; then
     k8s::resolveVersion "KUBERNETES_VERSION" "$KUBERNETES_VERSION"
     export KUBERNETES_VERSION=$resolveVersion
+
+    kind::prepareKindestImage $resolveVersion
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:50]:
```suggestion
    kind::prepareKindestImage "$resolveVersion"
```

File: scripts/ci-e2e-lib.sh [hunk 2]
```diff
@@ -40,27 +40,35 @@ capi:buildDockerImages () {
...
   fi
 
   if [ -n "${KUBERNETES_VERSION_UPGRADE_TO:-}" ]; then
     k8s::resolveVersion "KUBERNETES_VERSION_UPGRADE_TO" "$KUBERNETES_VERSION_UPGRADE_TO"
     export KUBERNETES_VERSION_UPGRADE_TO=$resolveVersion
+
+    kind::prepareKindestImage $resolveVersion
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:57]:
```suggestion
    kind::prepareKindestImage "$resolveVersion"
```

File: scripts/ci-e2e-lib.sh [hunk 3]
```diff
@@ -40,27 +40,35 @@ capi:buildDockerImages () {
...
   fi
 
   if [ -n "${KUBERNETES_VERSION_UPGRADE_FROM:-}" ]; then
     k8s::resolveVersion "KUBERNETES_VERSION_UPGRADE_FROM" "$KUBERNETES_VERSION_UPGRADE_FROM"
     export KUBERNETES_VERSION_UPGRADE_FROM=$resolveVersion
+
+    kind::prepareKindestImage $resolveVersion
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:64]:
```suggestion
    kind::prepareKindestImage "$resolveVersion"
```

File: scripts/ci-e2e-lib.sh [hunk 4]
```diff
@@ -40,27 +40,35 @@ capi:buildDockerImages () {
...
   fi
 
   if [ -n "${BUILD_NODE_IMAGE_TAG:-}" ]; then
     k8s::resolveVersion "BUILD_NODE_IMAGE_TAG" "$BUILD_NODE_IMAGE_TAG"
     export BUILD_NODE_IMAGE_TAG=$resolveVersion
+
+    kind::prepareKindestImage $resolveVersion
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:71]:
```suggestion
    kind::prepareKindestImage "$resolveVersion"
```

File: scripts/ci-e2e-lib.sh [hunk 5]
```diff
@@ -86,6 +94,84 @@ k8s::resolveVersion() {
...
+kind::prepareKindestImage() {
+  local version=$1
+
+  # Try to pre-pull the image
+  kind::prepullImage "kindest/node:$version"
+
+  # if pre-pull failed, falling back to local build
+  if [[ "$retVal" != 0 ]]; then
+    echo "+ image for Kuberentes $version is not available in docker hub, trying local build"
+    kind::buildNodeImage $version
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:108]:
```suggestion
    kind::buildNodeImage "$version"
```

File: scripts/ci-e2e-lib.sh [hunk 6]
```diff
@@ -86,6 +94,84 @@ k8s::resolveVersion() {
...
+  # sets the build version that will be applied by the Kubernetes build command called during kind build node-image.
+  k8s::setBuildVersion "$version"
+
+  # build the node image
+  version="${version//+/_}"
+  echo "+ Building kindest/node:$version"
+  kind build node-image --type docker --image "kindest/node:$version"
+
+  # move back to Cluster API
+  cd $REPO_ROOT
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:133]:
```suggestion
  cd $REPO_ROOT || exit
```

File: scripts/ci-e2e-lib.sh [hunk 7]
```diff
@@ -86,6 +94,84 @@ k8s::resolveVersion() {
...
+}
+
+# kind::buildNodeImage builds a kindest/node images starting from Kubernetes sources.
+# the func expect an input parameter defining the image tag to be used.
+kind::buildNodeImage() {
+  local version=$1
+
+  # move to the Kubernetes repository.
+  echo "KUBE_ROOT $GOPATH/src/k8s.io/kubernetes"
+  cd $GOPATH/src/k8s.io/kubernetes
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:119]:
```suggestion
  cd "$GOPATH"/src/k8s.io/kubernetes
```

File: scripts/ci-e2e-lib.sh [hunk 8]
```diff
@@ -86,6 +94,84 @@ k8s::resolveVersion() {
...
+
+  # checkout the required tag/branch.
+  local buildMetadata
+  buildMetadata=$(echo "${version#v}" | awk '{split($0,a,"+"); print a[2]}')
+  if [[ "$buildMetadata" == "" ]]; then
+    # if there are no release metadata, it means we are looking for a Kubernetes version that
+    # should be already been tagged.
+    echo "+ checkout tag $version"
+    git fetch --all --tags
+    git checkout tags/$version -b $version-branch
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:149]:
```suggestion
    git checkout tags/"$version" -b "$version-branch"
```

File: scripts/ci-e2e-lib.sh [hunk 9]
```diff
@@ -86,6 +94,84 @@ k8s::resolveVersion() {
...
+    echo "+ checking for existing branches"
+    git fetch --all
+
+    local major
+    local minor
+    major=$(echo "${version#v}" | awk '{split($0,a,"."); print a[1]}')
+    minor=$(echo "${version#v}" | awk '{split($0,a,"."); print a[2]}')
+
+    local releaseBranch
+    releaseBranch="$(git branch -r | grep release-$major.$minor$ || true)"
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:162]:
```suggestion
    releaseBranch="$(git branch -r | grep release-"$major"."$minor"$ || true)"
```

File: scripts/ci-e2e-lib.sh [hunk 10]
```diff
@@ -86,6 +94,84 @@ k8s::resolveVersion() {
...
+    if [[ "$releaseBranch" != "" ]]; then
+      # if there is already a release branch for the required Kubernetes branch, use it
+      echo "+ checkout $releaseBranch branch"
+      git checkout $releaseBranch -b release-$major.$minor
```
Comment (srm09) [inline scripts/ci-e2e-lib.sh:166]:
```suggestion
      git checkout $releaseBranch -b release-"$major"."$minor"
```

**Review by srm09 — COMMENTED** [review]
minor fixes for shell-check

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**